### ISSUE TITLE
Remove the piggieback dependency from figwheel

### ISSUE THREAD

**Issue by otijhuis** [bhauman/lein-figwheel#230]
Created a new issue regarding the removal the piggieback dependency. This was already partly discussed at the end of issue #180 .

**Comment by otijhuis** [comment]
@bhauman : like you mentioned I'll check to see if piggieback is available and otherwise run the regular cljs repl.

I'll also remove the defaults for the nrepl middleware and update the documentation accordingly.

**Comment by otijhuis** [comment]
Wondering about the best way to test this. In what situation would the piggieback repl normally be called? When starting it from within an nrepl connected session?

**Comment by bhauman** [comment]
Yep that's exactly when. Did you figure it out?

**Comment by otijhuis** [comment]
Was out salsa dancing and getting some sleep :). But yes, figured it out.

I managed to get it to work but there's a little problem. The call to the piggieback repl is still hardcoded in there. There's just no hard dependency anymore, it will check if it's available. This works fine with piggieback 0.1.5. But when I switch to 0.2.1 it no longer works. The function signature changed. The `:repl-env` option is gone and you just need to pass the env as the first parameter, just like cljs.repl. When I remove `:repl-env` it works with 0.2.1 but not with 0.1.5.

Basically, I can't leave anything hardcoded in there or there will be a "hard" dependency anyway. Looks like we'd need to pull out the code to start the piggieback repl completely if we want to remove the dependency.

Maybe the config.clj you mentioned is an option?

**Comment by bhauman** [comment]
I've been know to do a little salsa dancing myself.

Man thanks for working on this. This stuff is a pain.  Seems like there are no good default options.  Prt of its my fault for pulling this in the first place.  Should have just created a function that produced an env.

For now maybe we can do (apply piggieback/cljs-repl args) with a try catch to try the two apis??

**Comment by otijhuis** [comment]
Haha nice.

What's the best way to get the figwheel options to the `repl` function? Pass it along as param somewhere?

Something like this might work
```
             ;; Piggieback 0.1.5
             :nrepl-repl-fn (fn [figwheel-env opts]
                              (cemerick.piggieback/cljs-repl :repl-env figwheel-env :special-fns (or (:special-fns opts) cljs.repl/default-special-fns)))

             ;; Default
             ;;:nrepl-repl-fn cljs.repl/repl*

             ;; Piggieback 0.2.1
             ;; :nrepl-repl-fn (fn [figwheel-env opts]
             ;;                  (cemerick.piggieback/cljs-repl figwheel-env :special-fns (or (:special-fns opts) cljs.repl/default-special-fns))
```

**Comment by otijhuis** [comment]
Btw, is it possible to use the piggieback repl instead of the normal repl without nrepl since it passes everything on the cljs.repl? Then it might be possible to make the general repl configurable instead of having a specific version for nrepl alone.

**Comment by bhauman** [comment]
`cemerick.piggieback.rhino-repl-env` exists in piggieback 0.1.5 and doesn't exist in 0.2.1 we can use that to test which signature to call for now.

I really don't want to only go through piggieback.  It's very complex and its not the same as using the cljs.repl directly.  I like using the cljs.repl directly with inf-clojure just so that I know what is going on.

Thanks so much for your work on this. Each time I attempted it in the past it ate up lots of time and ended with a result that just didn't work well.

**Comment by otijhuis** [comment]
I didn't mean it that way. Was mostly curious about how piggieback works and reasoning about some stuff. I'd rather not deal with piggieback either if I don't have to.

If I can get to the figwheel-options in the `repl` function I think I can do the following

- Leave cljs.repl the default as is
- Completely remove every mention of piggieback from repl.clj
- Add an option like `:nrepl-repl-fn` to the config which takes a function with 2 args, the figwheel-env and the opts. Then it can anything it wants with those 2. Would make it possible to have a version for 0.1.5 and 0.2.1. Would just be a copy/paste action from the documentation for users. If an nREPL connection was found then the `repl` function can call the `:nrepl-repl-fn` function with the figwheel-env and the opts. In case `:nrepl-repl-fn` isn't defined it's set to `cljs.repl/repl*` which takes those 2 args anyway so no need for a wrapping fn.

```clojure
;; Piggieback 0.1.5
:nrepl-repl-fn (fn [figwheel-env opts]
                              (cemerick.piggieback/cljs-repl :repl-env figwheel-env :special-fns (or (:special-fns opts) cljs.repl/default-special-fns)))

;; Default if no option is given
;;:nrepl-repl-fn cljs.repl/repl*

;; Piggieback 0.2.1
;; :nrepl-repl-fn (fn [figwheel-env opts]
;;                  (cemerick.piggieback/cljs-repl figwheel-env :special-fns (or (:special-fns opts) cljs.repl/default-special-fns))
```

**Comment by bhauman** [comment]
I think I'm hearing you correctly, but this is such a limited communication medium.

I think this is a decent solution for now.  

Will this be able to pull in the figwheel special-functions?  

Will this work across the lein serialization boundary?  I'm a little worried about all the potential problems of arbitrary code not being able to execute because of the lein boundary, this could lead to lots of issues.

We could offer symbols `:piggieback-1` `:piggieback-2` and `:default` and generate the needed function.

We also may want to back off of this for now ... and see if a good idea comes to mind.  

I'm really not opposed to using `eval` and `resolve` to figure which environment we are in and offering up the correct default repl fn.

**Comment by otijhuis** [comment]
I'll just try it and see what works. I have some time so might as well give it a go. If it doesn't work I can always go back to `eval` / `resolve`. Don't know that much about leiningen. Only way to find out is trying it. Good way for me to learn anyway :)

Leiningen starts fine with the functions defined in the figwheel options. And the function would be called in the `repl` function by sidecar and get the parameters from there as well so I don't think there would be any issues.

Want to try something like this
```clojure
(defn repl
  ([build figwheel-server]
   (repl build figwheel-server {} {}))
  ([build figwheel-server opts]
   (repl build figwheel-server opts {}))
  ([build figwheel-server opts figwheel-options]
   (let [opts (merge (assoc (or (:compiler build) (:build-options build))
                            :warn-on-undeclared true)
                     opts)
         figwheel-repl-env (repl-env figwheel-server build)
         nrepl (thread-bound? #'nrepl-eval/*msg*)
         repl-opts (assoc opts :compiler-env (:compiler-env build))]
     (if nrepl
       (let [cljs-repl (or (:nrepl-repl-fn figwheel-options) #'cljs.repl/repl*)]
         (println "INFO: nREPL connection found. Using provided REPL or default if not provided")
         (cljs-repl figwheel-repl-env repl-opts))
       (do
         (println "INFO: No nREPL. Using default cljs REPL")
         (cljs.repl/repl* figwheel-repl-env repl-opts))))))
```

Is there any way to quickly get to the figwheel-options inside the `repl` fn without passing it along as an argument? Then I can quickly test it without modifying too much.

**Comment by otijhuis** [comment]
Have a couple of working scenario's.

**1) The original idea with the `:nrepl-repl-fn` defined in `project.clj`**

project.clj
```clojure
;; Piggieback 0.2.1
:nrepl-repl-fn (fn [figwheel-env opts]
                (cemerick.piggieback/cljs-repl figwheel-env :special-fns (or (:special-fns opts) cljs.repl/default-special-fns)))
```

Don't mind the `get-project-config` stuff. Just needed to get to the figwheel options quickly for testing.

```clojure
(defn repl
           ([build figwheel-server]
            (repl build figwheel-server {} (:figwheel (get-project-config))))
           ([build figwheel-server opts]
            (repl build figwheel-server opts (:figwheel (get-project-config))))
           ([build figwheel-server opts figwheel-options]
            (let [opts (merge (assoc (or (:compiler build) (:build-options build))
                                     :warn-on-undeclared true)
                              opts)
                  figwheel-repl-env (repl-env figwheel-server build)
                  nrepl (thread-bound? #'nrepl-eval/*msg*)
                  repl-opts (assoc opts :compiler-env (:compiler-env build))]
              (if nrepl
                (let [cljs-repl (or (eval (:nrepl-repl-fn figwheel-options)) #'cljs.repl/repl*)]
                  (println "INFO: nREPL connection found. Using provided REPL or default if not provided")
                  (cljs-repl figwheel-repl-env repl-opts))
                (do
                  (println "INFO: No nREPL. Using default cljs REPL")
                  (cljs.repl/repl* figwheel-repl-env repl-opts))))))
```

**2) Use a keyword like you suggested but in combination with a multimethod**

project.clj
```clojure
:nrepl-repl :piggieback
```

Multimethod
```clojure
(ns figwheel-sidecar.repls)

(defmulti start-cljs-repl (fn [nrepl repl figwheel-env opts]
                            [nrepl repl]))

(defmethod start-cljs-repl [true :piggieback]
  [nrepl _ figwheel-env opts]
  (try
    (require 'cemerick.piggieback)
    (let [cljs-repl (resolve 'cemerick.piggieback/cljs-repl)
          special-fns (or (:special-fns opts) cljs.repl/default-special-fns)]
      (try
        ;; Piggieback version >0.2
        (cljs-repl figwheel-env :special-fns special-fns)
        (println "Started piggieback 0.2+")
        (catch Exception e
          ;; Piggieback version 0.1.5
          (cljs-repl :repl-env figwheel-env :special-fns special-fns)
          (println "Started piggieback 0.1.5"))))
    (catch Exception e
      (println "WARNING: Unable to load piggieback. Check the project dependencies. Starting default REPL")
      (start-cljs-repl nrepl :default figwheel-env opts))))

(defmethod start-cljs-repl :default
  [_ _ figwheel-env opts]
  (println "INFO: Starting default cljs REPL")
  (cljs.repl/repl* figwheel-env opts))
```

`repl` fn
```clojure
(defn repl
  ([build figwheel-server]
   (repl build figwheel-server {} (:figwheel (get-project-config))))
  ([build figwheel-server opts]
   (repl build figwheel-server opts (:figwheel (get-project-config))))
  ([build figwheel-server opts figwheel-options]
   (let [opts (merge (assoc (or (:compiler build) (:build-options build))
                            :warn-on-undeclared true)
                     opts)
         figwheel-repl-env (repl-env figwheel-server build)
         repl-opts (assoc opts :compiler-env (:compiler-env build))
         nrepl (thread-bound? #'nrepl-eval/*msg*)
         nrepl-repl (or (:nrepl-repl figwheel-options) :default)]
     (repls/start-cljs-repl nrepl nrepl-repl figwheel-repl-env repl-opts))))
```

Quickly tested both with 0.1.5/0.2.1/default. Both versions seem to work fine in all cases but leaning towards the 2nd one. Easy to configure for the user like you suggested and yet it still allows someone to add a new `defmethod` or override an existing one.

**Comment by bhauman** [comment]
I'm traveling today.  But this looks good :)

**Comment by bhauman** [comment]
The second version :) but I'm wondering should we just check if we are in nrepl and then try to start piggieback if piggieback isn't there start the standard repl?  Why even have the  flag? Again there are only three possibilities?

Just wondering ...

**Comment by otijhuis** [comment]
Was thinking about the fact you might want to do things like running the default cljs repl inside a `lein repl` nrepl session where it would otherwise start a piggieback session by default if it's available.

But then again, without the flag that would still be possible with the multimethod implementation.

Figuring most stuff out while I go along so it's very likely I'm just overthinking things :)

Remove the flag, stick to the multimethod implementation? The `repl` fn stays very clean and you can still override stuff in the repl then. And very easy to implement as well.

**Comment by bhauman** [comment]
Man I have to say that I am grateful that you are taking the time to work on this.

I think the flag with the default detection would be perfect.  That way we get the best of both worlds.

If no flag is specified we can check if we are in an nrepl env ....

Again thanks for taking the time to work on this.  

I have another big push I'm going to do with figwheel to make it more amenable to scripting.
If you are interested in being involved in how that shapes up let me know.

**Comment by otijhuis** [comment]
Haha, it's nice to be able to contribute to something. I enjoy it and keeps me sharp as well :)

After your comments and a lot of thought, I think we can leave the flag out completely.

- if piggieback is a dependency and there's an nREPL connection the piggieback repl is started
- in all other cases the default cljs repl is started
- if you really want the piggieback dependency and still start the default cljs repl (or any other repl) you can always create your own `defmethod` and start the cljs repl afterwards

The first 2 seem to be the common use cases and the 3rd one is an edge case as far as I can tell. Without the flag the 3rd one is still possible, you just have to do it in a repl instead of the config. If someone is advanced enough to know the differences between the repls and things like that, that shouldn't be a problem. Can always put an example in the wiki.

Putting a flag in the config and passing the figwheel options to a lot of functions, just for an edge case, seems making things too complex without any obvious benefit.

Luckily one is almost an extension of the other. I'll implement it without the flag. If someone comes up with a valid reason for the flag I can always add it in later without having to refactor everything.

Should I put the `defmethod` stuff in another namespace or do you prefer having it in `repl.clj` with everything else.

And sure, if you think I can help out with something just let me know.

**Comment by bhauman** [comment]
Hey, I've been traveling and getting my stuff at home settled.

Yeah just keep it in the `repl.clj` for now. 

This is great!! Just send a pull request over and we will get it in here.

**Comment by otijhuis** [comment]
All the traveling I've been doing is going back and forth to work :)

Ok, I'll keep it there. The code is about done, just need to test it once more, see if I didn't forget anything.

I'll also update the docs, which is probably going to take the most time, and create the pull request.
Have the day off so should be done before the weekend.
